File: Python/sudoku.py
#!/usr/bin/env python -w
import sys


def row(pos):
    return pos/9


def col(pos):
    return pos % 9


def square(pos):
    row_g = row(pos)/3
    col_g = col(pos)/3
    return 10*row_g + col_g


def rcg(pos):
    return [row(pos), col(pos), square(pos)]

IDX = [ rcg(i) for i in range(0,81)]


def load_puzzle(fname):
    """Reads file and returns the array representation of the puzzle"""
    with open(fname,"r") as puzzle_file:
        puzzle_str = puzzle_file.read().replace('\n','')
        puzzle = list(puzzle_str)
    return puzzle


def expand_puzzles(puzzle_array):
    expanded_puzzles = [ ]
    pos = get_first_empty_position(puzzle_array[0])
    for puzzle in puzzle_array:
        expanded_puzzles += expand_puzzle(pos, puzzle)
    return expanded_puzzles


def get_first_empty_position(puzzle):
    return puzzle.index('_')


def replace_first(char, pos, puzzle):
    # this is the one place where we are not being functional, copying a value and replacing it in place.
    replaced = [] + puzzle
    replaced[pos] = char
    return replaced


def expand_puzzle(pos, puzzle):
    candidates = get_candidate_for_pos(pos, puzzle)
    return [ replace_first(x, pos, puzzle) for x in candidates ]


def get_candidate_for_pos(pos,puzzle):
    idx_puzzle = zip(IDX,puzzle)
    r,c,g = IDX[pos]
    vals =  [ x[1] for x in idx_puzzle if (x[0][0] == r) or (x[0][1] == c) or (x[0][2] == g) ]
    return set( ['1','2','3','4','5','6','7','8','9']) - set(vals)


def solved(puzzle):
    return puzzle.count('_') == 0


def solve( puzzle_array ):
    if solved(puzzle_array[0]):
        return puzzle_array
    print "%d puzzles" % len(puzzle_array)
    return solve( expand_puzzles(puzzle_array))


def print_puzzle( puzzle ):
    idx = 0
    for i in range(0,9):
        for j in range(0,9):
            print puzzle[idx],
            idx += 1
        print

def solve_puzzle(puzzle_name):
    puzzle = load_puzzle(puzzle_name)
    solutions = solve( [ puzzle ] )
    for solution in solutions:
        print_puzzle(solution)

if __name__ == "__main__":
    solve_puzzle(sys.argv[1])

File: Haskell/sudoku.hs
-- Sudoku Solver Challenge
--
-- Our solution is to keep a list of partial solutions, with each
-- iteration creating a new set of partial solutions by expanding
-- on the existing one, or removing it from the list of candidates
-- if no feasible solution can be derived from it.
import Data.List
import System.Environment

type Puzzle = [Char]

-- solve the puzzle
solve :: [Puzzle] -> Puzzle
solve [] = []   -- really should be raising an error
solve (x:xs)
  | solved x = x
  | otherwise = solve $ expand_puzzles (x:xs)

-- true if the puzzle is solved (i.e, no elements left with '_'
solved :: Puzzle -> Bool
solved []     = False
solved puzzle = '_' `notElem` puzzle

-- performs one iteration on solving the puzzle
expand_puzzles :: [Puzzle] -> [Puzzle]
expand_puzzles ps = flatmap expand_one_puzzle ps

-- expands one puzzle, creating a list of viable puzzles
-- derived from that one by replacing the first '_' with
-- the appropriate subset of '1' .. '9'
expand_one_puzzle :: Puzzle -> [Puzzle]
expand_one_puzzle puzzle =  map (add_to_puzzle puzzle) $ valid_candidates puzzle

-- adds the element to the puzzle, replacing the first '_'
add_to_puzzle :: Puzzle -> Char -> Puzzle
add_to_puzzle [] _ = []
add_to_puzzle ('_':xs) c = c:xs
add_to_puzzle (x:xs)   c = x : (add_to_puzzle xs c)

-- gets the subset of '1'..'9' that can be used to replace
-- the first '_' This involves finding the position for that
-- first '_', getting all the tiles for that group, column
-- and row and then doing a set difference to get only the
-- ones that are not defined. 
valid_candidates :: Puzzle -> [Char]
valid_candidates puzzle =
  let
    pos = position_of '_' puzzle
    row_t = row_tiles pos puzzle
    col_t = col_tiles pos puzzle
    grp_t = grp_tiles pos puzzle
  in
    "123456789" \\ (row_t ++ col_t ++ grp_t)

filter_by_first :: Int -> [(Int, Char)] -> [Char]
filter_by_first val zipped_puzzle = map snd $ filter (\v -> (fst v) == val) zipped_puzzle

-- returns the index of the first occurence of specified value in the array
position_of :: Char -> [Char] -> Int
position_of x arr =
  let
    arr_with_pos = zip [0..80] arr
    only_x = filter (\v -> (snd v) == x) arr_with_pos
  in
    if only_x == [] then -1 else fst (head only_x)

-- return the tiles belonging to the same row as position p
row_labels = map (`div` 9) [0..80]
row_tiles :: Int -> Puzzle -> [Char]
row_tiles pos puzzle = filter_by_first (row_labels !! pos) (zip row_labels puzzle)

-- returns the tiles belonging to the same columns as position p
col_labels = map (`mod` 9) [0..80]
col_tiles :: Int -> Puzzle -> [Char]
col_tiles pos puzzle = filter_by_first (col_labels !! pos) (zip col_labels puzzle)

-- returns the tiles belonging to the same group/square as position p
grp_func :: Int -> Int
grp_func pos =
  let
    r = row_labels !! pos
    c = col_labels !! pos
  in
    3 * (r `div` 3) + (c `div` 3)

grp_labels = map grp_func [0..80]
grp_tiles :: Int -> Puzzle -> [Char]
grp_tiles pos puzzle = filter_by_first (grp_labels !! pos) (zip grp_labels puzzle)

flatmap :: (a -> [b]) -> [a] -> [b]
flatmap func [] = []
flatmap func (x:xs) = func x ++ flatmap func xs

cleanup :: String -> Puzzle
cleanup str = filter (`elem` "123456789_") str

prettify :: Puzzle -> String
prettify puzzle = prettify_with_index $ zip [0..80] puzzle

prettify_with_index [] = "\n"
prettify_with_index (x:xs) 
  | (fst x) `mod` 9 == 0 = '\n' : (snd x) : (prettify_with_index xs)
  | otherwise            = (snd x) : (prettify_with_index xs)

main = do
    args <- getArgs
    input_puzzle <- readFile (head args)
    putStrLn input_puzzle
    putStrLn "---"
    putStrLn $  prettify $ solve [ ( cleanup input_puzzle ) ]
